### DISCUSSION TITLE
Testing with oath authentication

### DISCUSSION THREAD

**Discussion by marcelatoffernet** [fastapi/fastapi#6385] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import Depends, FastAPI, HTTPException, status

...

client = TestClient(app)

def test_foo(token: str = Depends(oauth2_scheme)):
    response = client.get("/foo")
    assert response.status_code == 200
    assert response.json() == {"Available Endpoints": ENDPOINT_LIST}
```

### Description

I followed the instructions for [oath2 with password](https://fastapi.tiangolo.com/tutorial/security/oauth2-jwt/) setup.

My API calls all work and are structured like so:

```python
@app.get("/foo")
async def read_foo(
        token: str = Depends(oauth2_scheme),
):
    return foo()
```

when I try to run the tests **in a different folder and file**, the authentication fails, ie i get a 401 response.

```python
from fastapi.testclient import TestClient

client = TestClient(app)

def test_foo(token: str = Depends(oauth2_scheme)):
    response = client.get("/foo")
    assert response.status_code == 200
    assert response.json() == {"Available Endpoints": ENDPOINT_LIST}
```

### Operating System

macOS

### FastAPI Version

0.75.2

### Python Version

Python 3.10.4

**Comment by jgould22** [comment]
Is your intention to run `test_foo` with pytest?

You cannot just toss a Depends in there and expect it to work. When you use `Depends` FastAPI knows what to do with it but pytest does not. To pytest you are just setting a default parameter. 

You have to authenticate with the test client as an api consumer would (sorta like [here](https://fastapi.tiangolo.com/tutorial/testing/?h=testing#extended-testing-file) ) or override the dependency like [here](https://fastapi.tiangolo.com/advanced/testing-dependencies/?h=over)

**Comment by marcelatoffernet** [comment, marked as answer]
@jgould22 Thanks! For anyone else with the same issue, I got it working by adding the Bearer token to the request:

```python
def test_get_endpoint_list():
    response = client.get("/", headers={
        "Authorization": f'Bearer {os.environ["BEARER_TOKEN"]}'})
    assert response.status_code == 200
    assert response.json() == {"Available Endpoints": ENDPOINT_LIST}

```

**Reply by FlorianPix** [reply]
Don't you need to request a token before the tests begin ?

**Reply by FlorianPix** [reply]
I found a way that works for me but it doesn't use the TestClient, instead I run the API in a docker container and then send requests from the tests to it. I did this because I got the authentication to work during testing but not the mongodb database. I always got None instead of the AsyncIOMotorClient.

### conftest.py

```python
testingUsers = {
    'test@test.de': 'test'
}

def getUserToken(user_name):
    #  client id and secret come from LogIn (Test Client)! which has password enabled under "Client > Advanced > Grant Types > Tick Password"
    url = 'your-auth-url'
    headers = {'content-type': 'application/json'}
    password = testingUsers[user_name]
    parameter = {"client_id": "your-client-id",
                 "client_secret": "your-client-secret",
                 "audience": 'your-audience',
                 "grant_type": "password",
                 "username": user_name,
                 "password": password, "scope": "openid"}
    #  do the equivalent of a CURL request from https://auth0.com/docs/quickstart/backend/python/02-using#obtaining-an-access-token-for-testing
    responseDICT = json.loads(requests.post(url, json=parameter, headers=headers).text)
    return responseDICT['access_token']

def getUserTokenHeaders(user_name='test@test.de'):
    return {'authorization': "Bearer " + getUserToken(user_name)}
```

### test_main.py

```python
def test_get_spot():
  url = 'http://localhost:8000/spot'
  headers = getUserTokenHeaders()
  response = requests.get(url, headers=headers)
  assert response.status_code == 200
  data = json.loads(response.text)
  assert data == []
```
